### DISCUSSION TITLE
Migration of the Epson package from version 3.1.1 to 4.1.1

### DISCUSSION THREAD

**Discussion by souhailtourjmen** [tr3v3r/react-native-esc-pos-printer#195] [Q&A] [answered]
Can you explain why there was a change in syntax between versions 3.1.1 and 4.1.1? What were the objectives or reasons behind this change? If we decide to stay with version 3.1.1, what would be the long-term impact, particularly in terms of support, security, compatibility with new features and performance?

**Comment by tr3v3r** [comment, marked as answer]
Hello @souhailtourjmen,
Here are the benefits:

1) Get rid of the `init` and `instantiate` methods and introduce a new class **Printer**
2)  **Printer** is as _singleton_ instance of the printer and can be used for single and multiple printing.
3) **Printer** class provides:
      - **print queue** on JS side
      - print SDK **methods** that are now just a **proxy from native SDK** ( the reason why API updated/changed)
      - **error handling** which includes description and native error code
      - proper **typing**
      

Hope it makes sense.

And sorry, I didn't write a migration guide because lack of time. But it is quite easy, just check the example folder

**Reply by souhailtourjmen** [reply]
Thank you for the explanation, it makes a lot of sense now!
